File: C++/BOJ/24479.cpp
#include<iostream>
#include<stack>
#include<vector>
#include<algorithm>

using namespace std;

int n, m, r;
vector<int> visited(0, n + 1);
vector<vector<int>> graph(m + 1, vector<int>({}));

void dfs()
{
    stack<int> stack;
    int tmp;
    int depth = 0;
    stack.push(r);
    while (stack.size())
    {
        tmp = stack.top();
        stack.pop();
        if (!visited[tmp])
        {   
            depth += 1;
            visited[tmp] = depth;
            for (auto num : graph[tmp])
                stack.push(num);
        }
    }
}

int main()
{
    cin.tie(0);
	cout.tie(0);
	ios::sync_with_stdio(0);

    int left, right;
    cin>>n>>m>>r;
    for (int i = 0; i < m; i++)
    {
        cin>>left>>right;
        graph[left].push_back(right);
        graph[right].push_back(left);
    }
    for (int i = 1; i < m + 1; i++)
        sort(graph[i].begin(), graph[i].end(), greater<int>());
    dfs();
    for (int i = 1; i < m + 1; i++)
        cout<<visited[i]<<"\n";
}

File: Python/BOJ/24479.py
import sys
n, m, r = map(int, sys.stdin.readline().split())

graph = [[] for _ in range(n + 1)]
visited=[0 for _ in range(n + 1)]
for i in range(1, m+1):
    left, right = map(int, sys.stdin.readline().split())
    graph[left].append(right)
    graph[right].append(left)

for i in range(1, n + 1):
    graph[i] = sorted(graph[i], reverse= True)

def dfs():
    depth = 0
    stack = []
    stack.append(r)
    while stack:
        tmp = stack.pop()
        if not visited[tmp]:
            depth += 1
            visited[tmp] = depth
            for num in graph[tmp]:
                stack.append(num)

dfs()
for v in visited[1:]:
    print(v)